### ISSUE TITLE
Add support for native Go implementation of AsciiDoc

### ISSUE THREAD

**Issue by anthonyfok** [spf13/hugo#1435]
Starting with v0.14 (released in May 2015), Hugo provides preliminary support for AsciiDoc with the help of an external helper such as `asciidoc` (Python-base) and `asciidoctor` (Ruby-based).

Please see Issue #470 "Support for AsciiDoc" (which led to the current experimental "external helper" implementation) for more information.  This issue 

However, the ultimate goal is for Hugo to support a native Go implementation of AsciiDoc, something that e.g. the https://github.com/VonC/asciidocgo project set out to accomplish.  This issue will track the progress on that front.

**Comment by bguiz** [comment]
Note that the linked AsciiDocGo project is unmaintained, and is likely to remain so - see https://github.com/VonC/asciidocgo/issues/1#issuecomment-72535213 - where the author says it only covers about 10-15% of the surface area covered by AsciiDoctor.

That being said, I would still very much like to see this happen. This begs the question - is Hugo going to attempt to use AsciiDocGo, or write its own implementation?

**Comment by anthonyfok** [comment]
Yes, @bguiz, I agree with you totally.  Consider this issue more a "Call for Action" for volunteers to take up this worthy but challenging task.  It won't happen right away, but if the demand is high enough, I am sure someone will take up the task and either continue with the great work that @VonC started, or start a new implementation if that is easier.  Perhaps some kind of open-source bounty or crowdfunding could be considered too, a possibility as our community continue to grow.  Just some random ideas.  :-)

Meanwhile, I am starting to think that whether some kind of a mythical Ruby-to-Go converter would be feasible to help with this task, haha!  I can always dream.  :wink:

**Comment by kastork** [comment]
I wonder if there might be a way to use https://github.com/robertkrimen/otto to render asciidoc with
https://github.com/asciidoctor/asciidoctor.js

**Comment by bep** [comment]
That sounds slow.

**Comment by kastork** [comment]
Could be.  If the otto is slow then it is sure to be.  But it might be a way to make the asciidoc solution self-contained rather than relying on the external process (which is also slow).

Anyway, just tossing it out there.  I was inspired by the mythical ruby-to-go converter comment above. (asciidoc.js is generated by a ruby-to-js converter from what I understand).

**Comment by pgorod** [comment]
I bought into the Asciidoc thing and invested in it for my new Hugo site. Only now as the content grows I begin to realize the performance penalty I'm getting from it. The projects linked here seem completely stalled, unfortunately...

I'm using the Ruby Asciidoctor. Is the Python-based Asciidoc any less horrible in terms of performance?

It's a real pity, because apart from the slow compilations, I'm really loving the format... a perfect fit for Hugo if you ask me - modern, well-designed, simple and direct.

**Comment by raghur** [comment]
THere's a new library coming up - https://github.com/bytesparadise/libasciidoc It's probably still early days but it's driven off the PEG grammar so much likely to be robust. 

I'm posting this here in the hope that folks who're interested and invested in asciidoc can probably head out to the libasciidoc project and help out.

**Comment by etsauer** [comment]
@xcoulon have you made any progress with `libasciidoc`?

**Comment by xcoulon** [comment]
@etsauer sorry for the delayed response. I'm still working on libasciidoc, indeed, and there are still a [handful of issues](https://github.com/bytesparadise/libasciidoc/issues?q=is%3Aopen+is%3Aissue+milestone%3Av0.1) that I'd like to address on before making a `v0.1` release. 
I was able to integrate libasciidoc as a rendering engine in Hugo a while ago, but it would probably need a little more work before I can propose a PR. One thing I had in mind was adding it as a configurable, experimental feature so users can still rely on Asciidoctor if they prefer.  Once `v0.1` is released, I'll definitely get back to the Hugo integration.

**Comment by etsauer** [comment]
@xcoulon yes, a feature toggle for the first release would be nice. Let me know when there is something worth testing.

**Comment by cup** [comment]
https://github.com/gohugoio/hugo/pull/5672

**Comment by shuLhan** [comment]
Shameless plug here, sorry, there is another asciidoc library [asciidoctor-go](https://sr.ht/~shulhan/asciidoctor-go/) in case anyone interested and have time to integrate it with hugo.

Proof of concept: [here](https://git.sr.ht/~shulhan/asciidoctor-go/tree/main/item/testdata), [here](https://git.sr.ht/~shulhan/kilabit.info/tree/main/item/_content) generate this [site](https://kilabit.info), and [here](https://github.com/golang-id/web/tree/master/_content) generate this [site](https://golang-id.org).

**Comment by xcoulon** [comment]
Nice work @shuLhan! I'm still make progress on [libasciidoc](https://github.com/bytesparadise/libasciidoc) and I hope I can start working on the integration with Hugo soon now 🤞
